### PR TITLE
IBM i support

### PR DESCRIPTION
**Pull request by dmabupt**

### CODE REVIEW

File: src/unix/ibmi.c [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
```
Comment (gireeshpunathil) [inline src/unix/ibmi.c]: I believe new files need not carry old copyright notices.

File: src/unix/ibmi.c [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+void uv_loadavg(double avg[3]) {
+    avg[0] = 0.; avg[1] = 0.; avg[2] = 0.;
+    return;
+}
+
+int uv_resident_set_memory(size_t* rss) {
+  /* FIXME: read /proc/meminfo? */
+  *rss = 0;
+  return UV_ENOSYS;
```
Comment (gireeshpunathil) [inline src/unix/ibmi.c]: If we are returning `no system support`, we don't need to modify rss right? we can leave it untouched?

File: src/unix/ibmi.c [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+
+void uv_loadavg(double avg[3]) {
+    avg[0] = 0.; avg[1] = 0.; avg[2] = 0.;
+    return;
+}
+
+int uv_resident_set_memory(size_t* rss) {
+  /* FIXME: read /proc/meminfo? */
```
Comment (gireeshpunathil) [inline src/unix/ibmi.c]: Is this a work in progress, are there ways to get the RSS through system calls?

File: src/unix/ibmi.c [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    cpu_info->speed = 0;
+    cpu_info->model = uv__strdup("unknown");
+    cpu_info->cpu_times.user = 0;
+    cpu_info->cpu_times.sys = 0;
+    cpu_info->cpu_times.idle = 0;
+    cpu_info->cpu_times.irq = 0;
+    cpu_info->cpu_times.nice = 0;
+    cpu_info++;
+    idx++;
+  }
```
Comment (gireeshpunathil) [inline src/unix/ibmi.c]: Looks like all what we get from system is the number of CPUs? If so, do we need to run through each fields of each entries in the structure, can't we just `memset` the  allocated memory?

File: uv.gyp [hunk 5]
```diff
@@ -274,19 +274,44 @@
...
+                'src/unix/posix-poll.c',
+                'src/unix/no-fsevents.c',
+                'src/unix/no-proctitle.c',
+              ],
+              'defines': [
+                '_PASE'
+              ],
+            }],
+            [ '"<(TrueOS)"!="OS400"', {
+              'sources': [
```
Comment (gireeshpunathil) [inline uv.gyp]: Instead of checking for inequality with `OS400`, can't we do equality check with the expected name?

File: src/unix/aix-ibmi.c [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+void uv_free_interface_addresses(uv_interface_address_t* addresses,
+  int count) {
+  int i;
+
+  for (i = 0; i < count; ++i) {
+    uv__free(addresses[i].name);
+  }
+
+  uv__free(addresses);
+}
```
Comment (gireeshpunathil) [inline src/unix/aix-ibmi.c]: some special character here?

**Comment by cjihrig** [comment]
Honest question: how many IBM platforms are we likely going to end up supporting? IBM has been doing all the work supporting them, so it's not necessarily a bad thing. I'm just curious.

**Comment by gireeshpunathil** [comment]
If I understand correctly:
```
AIX on PPC64 BE
Linux on PPC64 BE
Linux on PPC64 LE
Linux on S390x
IBM i (this one)
```

File: checksparse.sh [hunk 7]
```diff
@@ -177,8 +177,18 @@ case `uname -s` in
 AIX)
   SPARSE_FLAGS="$SPARSE_FLAGS -D_AIX=1"
   SOURCES="$SOURCES
+           src/unix/aix-ibmi.c
            src/unix/aix.c"
   ;;
+OS400)
+  SPARSE_FLAGS="$SPARSE_FLAGS -D_AIX=1"
```
Comment (bnoordhuis) [inline checksparse.sh]: `-D_PASE=1`?

File: src/unix/aix-ibmi.c [hunk 8]
```diff
@@ -0,0 +1,292 @@
+/* Copyright Joyent, Inc. and other Node contributors. All rights reserved.
```
Comment (bnoordhuis) [inline src/unix/aix-ibmi.c]: New files should use `Copyright libuv project contributors. All rights reserved.`

**Comment by cjihrig** [comment]
Thanks @gireeshpunathil!

File: src/unix/aix-ibmi.c [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+    memcpy(buffer, abspath, *size);
+    buffer[*size] = '\0';
+
+    return 0;
+  } else {
+  /* Case iii). Search PATH environment variable */
+    char trypath[PATH_MAX];
+    char *clonedpath = NULL;
+    char *token = NULL;
+    char *path = getenv("PATH");
```
Comment (bnoordhuis) [inline src/unix/aix-ibmi.c]: `getenv()` and `strtok()` are not MT-safe.  Is there another way to do this?

(I realize this comes from aix.c but since we're here.)

File: src/unix/aix-ibmi.c [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+  int* count) {
+  uv_interface_address_t* address;
+  int sockfd, inet6, size = 1;
+  struct ifconf ifc;
+  struct ifreq *ifr, *p, flg;
+  struct sockaddr_dl* sa_addr;
+
+  *count = 0;
+
+  if (0 > (sockfd = socket(AF_INET, SOCK_DGRAM, IPPROTO_IP))) {
```
Comment (bnoordhuis) [inline src/unix/aix-ibmi.c]: I realize this is copied from aix.c but can you split out the assignment from the comparison?  That style isn't normally used in the libuv code base.

As well, it would be better to use `uv__socket()`, it marks the socket close-on-exec (relevant when another thread forks while this code is executing.)

File: src/unix/ibmi.c [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+
+uint64_t uv_get_total_memory(void) {
+  return (uint64_t) sysconf(_SC_PAGESIZE) * sysconf(_SC_PHYS_PAGES);
+}
+
+
+void uv_loadavg(double avg[3]) {
+    avg[0] = 0.; avg[1] = 0.; avg[2] = 0.;
```
Comment (bnoordhuis) [inline src/unix/ibmi.c]: A shorter way of writing that is `avg[0] = avg[1] = avg[2] = 0;`

File: src/unix/ibmi.c (hunk 2 above)
Comment (bnoordhuis) [reply src/unix/ibmi.c]: I speculate this code comes from cygwin.c but yes, `rss` doesn't need (and arguably shouldn't) be touched on error.  `uv_uptime()` three lines below doesn't do that either.

File: src/unix/ibmi.c [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+  uv_cpu_info_t* cpu_info;
+
+  *cpu_infos = NULL;
+  *count = 0;
+
+  numcpus = sysconf(_SC_NPROCESSORS_ONLN);
+  assert(numcpus != (unsigned int) -1);
+  assert(numcpus != 0);
+
+  *cpu_infos = (uv_cpu_info_t*) uv__malloc(numcpus * sizeof(uv_cpu_info_t));
```
Comment (bnoordhuis) [inline src/unix/ibmi.c]: Superfluous cast.

File: src/unix/ibmi.c [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+  if (!*cpu_infos) {
+    return -ENOMEM;
+  }
+
+  cpu_info = *cpu_infos;
+  while (idx < numcpus) {
```
Comment (bnoordhuis) [inline src/unix/ibmi.c]: `for (idx = 0; idx < numcpus; idx++) {`?

File: src/unix/ibmi.c (hunk 4 above)
Comment (bnoordhuis) [reply src/unix/ibmi.c]: Shouldn't really matter performance-wise and you'd still have to do another pass to set the `model` field.

File: uv.gyp [hunk 14]
```diff
@@ -274,19 +274,44 @@
           },
         }],
         [ 'OS=="aix"', {
-          'sources': [ 'src/unix/aix.c' ],
+          'variables': {
+            'TrueOS': '<!(uname -s)',
```
Comment (bnoordhuis) [inline uv.gyp]: Suggestion: name this `uname_s` or `os_name` or something?

File: uv.gyp [hunk 15]
```diff
@@ -274,19 +274,44 @@
...
+          'conditions': [
+            [ '"<(TrueOS)"=="OS400"', {
+              'sources': [ 
+                'src/unix/ibmi.c',
+                'src/unix/posix-poll.c',
+                'src/unix/no-fsevents.c',
+                'src/unix/no-proctitle.c',
+              ],
+              'defines': [
+                '_PASE'
```
Comment (bnoordhuis) [inline uv.gyp]: I'm curious, isn't that set automatically by xlc?  (And gcc too, presumably?)

File: uv.gyp (hunk 5 above)
Comment (bnoordhuis) [reply uv.gyp]: That or fold it into the previous condition like this:
```python
[ '"<(TrueOS)"=="OS400"', {
  # ...
}, {
  # ...
}],
```

File: src/unix/aix-ibmi.c (hunk 6 above)
Comment (dmabupt) [reply src/unix/aix-ibmi.c]: I could not find any special character here.

File: checksparse.sh (hunk 7 above)
Comment (dmabupt) [reply checksparse.sh]: It should be -D_PASE=1

File: uv.gyp (hunk 15 above)
Comment (dmabupt) [reply uv.gyp]: gcc is the compiler we used and it would not set this.

File: src/unix/ibmi.c (hunk 3 above)
Comment (dmabupt) [reply src/unix/ibmi.c]: It is very difficult to get memory related info on IBM i. I suggest to bypass it for now.

File: uv.gyp (hunk 14 above)
Comment (dmabupt) [reply uv.gyp]: I think os_name is better.

File: src/unix/aix-ibmi.c (hunk 6 above)
Comment (gireeshpunathil) [reply src/unix/aix-ibmi.c]: I am not too sure what it represents, but when I open the changes through github and `Files changed` tab and come to the bottom of this file, I am seeing this one:
special

Can you open the file in binary and check if the value of last byte is `125` (`}`) or something else?

File: src/unix/ibmi.c (hunk 3 above)
Comment (gireeshpunathil) [reply src/unix/ibmi.c]: sure, no problem

File: src/unix/ibmi.c (hunk 1 above)
Comment (dmabupt) [reply src/unix/ibmi.c]: Updated with new copyright text.

File: src/unix/ibmi.c (hunk 2 above)
Comment (dmabupt) [reply src/unix/ibmi.c]: Updated the code to keep rss untouched.

File: uv.gyp (hunk 5 above)
Comment (dmabupt) [reply uv.gyp]: The folded style is better. Updated the code.

File: src/unix/aix-ibmi.c (hunk 6 above)
Comment (dmabupt) [reply src/unix/aix-ibmi.c]: Yes, it is 0x7D which is }

File: checksparse.sh (hunk 7 above)
Comment (dmabupt) [reply checksparse.sh]: Updated the code

File: src/unix/aix-ibmi.c (hunk 8 above)
Comment (dmabupt) [reply src/unix/aix-ibmi.c]: Updated the copyright text

File: src/unix/ibmi.c (hunk 11 above)
Comment (dmabupt) [reply src/unix/ibmi.c]: Updated the code

File: src/unix/ibmi.c (hunk 12 above)
Comment (dmabupt) [reply src/unix/ibmi.c]: Updated the code to remove these lines.

File: src/unix/ibmi.c (hunk 13 above)
Comment (dmabupt) [reply src/unix/ibmi.c]: Updated the code.

File: uv.gyp (hunk 14 above)
Comment (dmabupt) [reply uv.gyp]: Updated the code.

File: checksparse.sh [hunk 16]
```diff
@@ -181,7 +181,7 @@ AIX)
            src/unix/aix.c"
   ;;
 OS400)
-  SPARSE_FLAGS="$SPARSE_FLAGS -D_AIX=1"
+  SPARSE_FLAGS="$SPARSE_FLAGS -D_PASE=1"
```
Comment (dmabupt) [inline checksparse.sh]: Change to use _PASE

File: src/unix/aix.c [hunk 17]
```diff
@@ -1,4 +1,5 @@
-/* Copyright Joyent, Inc. and other Node contributors. All rights reserved.
+/* Copyright libuv project contributors. All rights reserved.
```
Comment (dmabupt) [inline src/unix/aix.c]: Update copyright text

File: src/unix/ibmi.c [hunk 18]
```diff
@@ -1,4 +1,5 @@
-/* Copyright Joyent, Inc. and other Node contributors. All rights reserved.
+/* Copyright libuv project contributors. All rights reserved.
```
Comment (dmabupt) [inline src/unix/ibmi.c]: Update copyright text

File: src/unix/ibmi.c [hunk 19]
```diff
@@ -66,13 +67,11 @@ uint64_t uv_get_total_memory(void) {
 
 
 void uv_loadavg(double avg[3]) {
-    avg[0] = 0.; avg[1] = 0.; avg[2] = 0.;
+    avg[0] = avg[1] = avg[2] = 0;
```
Comment (dmabupt) [inline src/unix/ibmi.c]: Make code more compact

File: src/unix/ibmi.c [hunk 20]
```diff
@@ -66,13 +67,11 @@ uint64_t uv_get_total_memory(void) {
...
     return;
 }
 
 int uv_resident_set_memory(size_t* rss) {
-  /* FIXME: read /proc/meminfo? */
-  *rss = 0;
```
Comment (dmabupt) [inline src/unix/ibmi.c]: Keep rss untouched

File: src/unix/ibmi.c [hunk 21]
```diff
@@ -89,16 +88,14 @@ int uv_cpu_info(uv_cpu_info_t** cpu_infos, int* count) {
   *count = 0;
 
   numcpus = sysconf(_SC_NPROCESSORS_ONLN);
-  assert(numcpus != (unsigned int) -1);
```
Comment (dmabupt) [inline src/unix/ibmi.c]: Remove redundant code

File: src/unix/ibmi.c [hunk 22]
```diff
@@ -89,16 +88,14 @@ int uv_cpu_info(uv_cpu_info_t** cpu_infos, int* count) {
...
-  assert(numcpus != 0);
 
   *cpu_infos = (uv_cpu_info_t*) uv__malloc(numcpus * sizeof(uv_cpu_info_t));
   if (!*cpu_infos) {
     return -ENOMEM;
   }
 
   cpu_info = *cpu_infos;
-  while (idx < numcpus) {
+  for (idx = 0; idx < numcpus; idx++) {
```
Comment (dmabupt) [inline src/unix/ibmi.c]: Updated the code

File: uv.gyp [hunk 23]
```diff
@@ -275,7 +275,7 @@
         }],
         [ 'OS=="aix"', {
           'variables': {
-            'TrueOS': '<!(uname -s)',
+            'os_name': '<!(uname -s)',
```
Comment (dmabupt) [inline uv.gyp]: Change the true os name var to 'os_name'

File: uv.gyp [hunk 24]
```diff
@@ -297,8 +297,7 @@
               'defines': [
                 '_PASE'
               ],
-            }],
-            [ '"<(TrueOS)"!="OS400"', {
+            }, {
```
Comment (dmabupt) [inline uv.gyp]: Make the code more compact

File: src/unix/aix-ibmi.c (hunk 6 above)
Comment (gireeshpunathil) [reply src/unix/aix-ibmi.c]: ok, so probably it is just a `github` quirck. Let us forget it.

**Comment by mhdawson** [comment]
I think that these files are as follows:

src/unix/aix-ibmi.c  - only for aix, not OS400
src/unix/ibmi.c - only for OS400, not aix
src/unix/aix.c - common to both OS400 and aix

I don't find that the names lead you to this understanding.  How about:

src/unix/aix-only.c  - only for aix, not OS400
src/unix/ibmi-only.c - only for OS400, not aix
src/unix/aix.c - common to both OS400 and aix

or
src/unix/aix-only.c  - only for aix, not OS400
src/unix/ibmi-only.c - only for OS400, not aix
src/unix/aix-common.c - common to both OS400 and aix

Open to other suggestions as well.

File: src/unix/ibmi.c [hunk 25]
```diff
@@ -0,0 +1,111 @@
...
+
+
+void uv_loadavg(double avg[3]) {
+    avg[0] = avg[1] = avg[2] = 0;
+    return;
+}
+
+int uv_resident_set_memory(size_t* rss) {
+  return UV_ENOSYS;
+
```
Comment (jBarz) [inline src/unix/ibmi.c]: Unnecessary blank line

File: src/unix/ibmi.c [hunk 26]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+int uv_uptime(double* uptime) {
+  return UV_ENOSYS;
+}
+
```
Comment (jBarz) [inline src/unix/ibmi.c]: I think the style convention is to have 2 blank lines between functions

File: uv.gyp (hunk 15 above)
Comment (jBarz) [reply uv.gyp]: Given that there are more differences than commonalities in this section between aix and OS400, I think the os400 section should be separate from aix maybe!
```gyp
[ 'OS=="aix"', {
   ...
}],
[ 'OS=="os400"', {
   ...
}],
```

File: uv.gyp [hunk 27]
```diff
@@ -274,19 +274,43 @@
           },
         }],
         [ 'OS=="aix"', {
-          'sources': [ 'src/unix/aix.c' ],
+          'variables': {
```
Comment (jBarz) [inline uv.gyp]: Visually, there seem to be more differences than commonalities in this section between aix and OS400, I think the os400 section should be separate from aix maybe!
```gyp
[ 'OS=="aix"', {
   ...
}],
[ 'OS=="os400"', {
   ...
}],
```

Comment (dmabupt) [reply uv.gyp]: Currently the IBM i python returns "aix6" for sys.platform. It's great when everything in PASE works the same in AIX (the normal conditions), but in some rare conditions (like libuv), we have to handle it specifically.

File: src/unix/ibmi.c (hunk 26 above)
Comment (dmabupt) [reply src/unix/ibmi.c]: Commit [46ecf1c](https://github.com/libuv/libuv/pull/1601/commits/46ecf1c7ad270660b9c2103a584dd67bdda08cdf) updated the style.

File: src/unix/ibmi.c (hunk 25 above)
Comment (dmabupt) [reply src/unix/ibmi.c]: Commit [46ecf1c](https://github.com/libuv/libuv/pull/1601/commits/46ecf1c7ad270660b9c2103a584dd67bdda08cdf) updated the style.

File: uv.gyp (hunk 15 above)
Comment (dmabupt) [reply uv.gyp]: No. Maybe in future we could build an IBM i gcc which can handle this. But for now, we use the AIX gcc which does not automatically add this macro.

**Comment by dmabupt** [comment]
@mhdawson , for the new file names, how about the shorter ones --
src/unix/**aix**.c - only for aix, not OS400
src/unix/**ibmi**.c - only for OS400, not aix
src/unix/**aix-common**.c - common to both OS400 and aix

File: src/unix/aix-ibmi.c [hunk 28]
```diff
@@ -0,0 +1,293 @@
...
+  return (uint64_t) t.tb_high * G + t.tb_low;
+}
+
+
+/*
+ * We could use a static buffer for the path manipulations that we need outside
+ * of the function, but this function could be called by multiple consumers and
+ * we don't want to potentially create a race condition in the use of snprintf.
+ * There is no direct way of getting the exe path in AIX - either through /procfs
+ * or through some libc APIs. The below approach is to parse the argv[0]'s pattern
```
Comment (bnoordhuis) [inline src/unix/aix-ibmi.c]: Can you check your changes for lines > 80 columns?

File: src/unix/aix-ibmi.c [hunk 29]
```diff
@@ -0,0 +1,293 @@
...
+    *size -= 1;
+    if (*size > abspath_size)
+      *size = abspath_size;
+
+    memcpy(buffer, abspath, *size);
+    buffer[*size] = '\0';
+
+    return 0;
+  } else {
+  /* Case iii). Search PATH environment variable */
```
Comment (bnoordhuis) [inline src/unix/aix-ibmi.c]: Can you indent this by two more columns?

File: src/unix/aix-ibmi.c [hunk 30]
```diff
@@ -0,0 +1,293 @@
...
+  }
+
+  ifc.ifc_req = (struct ifreq*)uv__malloc(size);
+  ifc.ifc_len = size;
+  if (ioctl(sockfd, SIOCGIFCONF, &ifc) == -1) {
+    uv__close(sockfd);
+    return -errno;
+  }
+
+#define ADDR_SIZE(p) MAX((p).sa_len, sizeof(p))
```
Comment (bnoordhuis) [inline src/unix/aix-ibmi.c]: Is MAX from a system header?

File: src/unix/aix-ibmi.c (hunk 6 above)
Comment (bnoordhuis) [reply src/unix/aix-ibmi.c]: I think what you're seeing is GitHub pointing out that there is no newline at the end of the file.

File: src/unix/aix.c (hunk 17 above)
Comment (bnoordhuis) [reply src/unix/aix.c]: Can you revert that?  This is an old file with the old copyright.

File: uv.gyp (hunk 15 above)
Comment (bnoordhuis) [reply uv.gyp]: Can you define it as `_PASE=1` for consistency?

File: src/unix/aix-ibmi.c (hunk 6 above)
Comment (gireeshpunathil) [reply src/unix/aix-ibmi.c]: ok, thanks!

**Comment by mhdawson** [comment]
@dmabupt your suggested file names are good with me.

@bnoordhuis 'libuv+zos CI'  -> yes.  Its here https://ci.nodejs.org/job/libuv-test-commit-zos/.  @jBarz is working on the last failure before I add it to the regular libuv runs.   **But** that is zOS not system i which is different.  I've talk to the system i lead about hardware so its on his radar but we've not figure out how to make that happen for system it yet.

File: src/unix/aix-ibmi.c (hunk 28 above)
Comment (dmabupt) [reply src/unix/aix-ibmi.c]: The max line length in this part is 80 (not including the \n character). Is that OK?

File: src/unix/aix-ibmi.c (hunk 30 above)
Comment (dmabupt) [reply src/unix/aix-ibmi.c]: No warning reported during compiling this file. Guees it is defined in some system header.

File: src/unix/aix-ibmi.c (hunk 29 above)
Comment (dmabupt) [reply src/unix/aix-ibmi.c]: Commit [c3716ea](https://github.com/libuv/libuv/pull/1601/commits/c3716ea34fdc8f1cbea2bbe23c27389cf8128a15) updated the code

**Comment by dmabupt** [comment]
Found a node inspector issue when I was testing the new code with Node.js v8.8.0. 
Seems the **inspector_io.cc** file does not include the correct uv header files so the structure **uv_loop_t** is different from the one in **posix-poll.c** of libuv, which leads to core dump in PASE.  
Is this a node inspector issue?

- inspector_io.cc--InspectorIo::ThreadMain():  sizeof(uv_loop_t)=824;
- posix-poll.c--uv__io_poll(): sizeof(uv_loop_t)=856

**Comment by mhdawson** [comment]
@bnoordhuis can you take a look to see if all of your comments have been addressed.

File: src/unix/aix-common.c [hunk 31]
```diff
@@ -0,0 +1,293 @@
...
+
+    if (!(flg.ifr_flags & IFF_UP && flg.ifr_flags & IFF_RUNNING))
+      continue;
+
+    (*count)++;
+  }
+
+  /* Alloc the return interface structs */
+  *addresses = (uv_interface_address_t*)
+    uv__malloc(*count * sizeof(uv_interface_address_t));
```
Comment (bnoordhuis) [inline src/unix/aix-common.c]: Unnecessary cast.  (I realize this is moved over from another file but since you're here.)

File: src/unix/aix.c [hunk 32]
```diff
@@ -1,4 +1,5 @@
 /* Copyright Joyent, Inc. and other Node contributors. All rights reserved.
+ *
```
Comment (bnoordhuis) [inline src/unix/aix.c]: Can you undo this change?

File: src/unix/ibmi.c [hunk 33]
```diff
@@ -0,0 +1,112 @@
+/* Copyright Joyent, Inc. and other Node contributors. All rights reserved.
```
Comment (bnoordhuis) [inline src/unix/ibmi.c]: New file, old copyright.  Can you update?

File: src/unix/aix.c (hunk 17 above)
Comment (bnoordhuis) [reply src/unix/aix.c]: I don't know if you accidentally unreverted this but the old copyright should stay in this file.

**Comment by mhdawson** [comment]
CI run: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/557/

@gireeshpunathil , @jBarz any other comments before this lands ?

**Comment by gireeshpunathil** [comment]
code changes look good. I guess some clarity on copyright will be good:
 - aix.c : old file (need to retain the joyent copyright, currently it is not)
 - aix-common.c: new file, but its predominent content belongs to old aix.c - so not sure what copyright applies here.
 - ibmi.c - fresh file, currently contain copyright on libuv project owners - is this a mandate? or it could carry the copyright, if any, from its original owner (@dmabupt)?

**Comment by dmabupt** [comment]
I have squashed the commits. Any comments on the description?

File: src/unix/aix.c (hunk 17 above)
Comment (bnoordhuis) [reply src/unix/aix.c]: This is still unfixed, it looks like...

File: src/unix/ibmi.c [hunk 34]
```diff
@@ -0,0 +1,112 @@
...
+int uv_cpu_info(uv_cpu_info_t** cpu_infos, int* count) {
+  unsigned int numcpus, idx = 0;
+  uv_cpu_info_t* cpu_info;
+
+  *cpu_infos = NULL;
+  *count = 0;
+
+  numcpus = sysconf(_SC_NPROCESSORS_ONLN);
+
+  *cpu_infos = (uv_cpu_info_t*) uv__malloc(numcpus * sizeof(uv_cpu_info_t));
```
Comment (bnoordhuis) [inline src/unix/ibmi.c]: Unnecessary cast.

**Comment by dmabupt** [comment]
I was confused a little bit for the copyright changes. Seems only the new files (aix-common.c & ibmi.c) should use the new copyright header ("libuv...") while the old file (aix.c) keep the old header ("Joyent"), right?
I reverted the aix.c's copyright to "Joyent" again via commit dd9b242. Is that correct?

**Comment by dmabupt** [comment]
Updated the commit description. Any comments?

**Comment by bnoordhuis** [comment]
@dmabupt https://ci.nodejs.org/job/libuv-test-commit-aix/567/nodes=aix61-ppc64/console - can you fix the build breakage?

**Comment by mhdawson** [comment]
Looks like CI is no longer visible, new ci run https://ci.nodejs.org/view/libuv/job/libuv-test-commit/574/

**Comment by bnoordhuis** [comment]
For posterity:
<details>

```
+ make -j2
  CC       src/libuv_la-fs-poll.lo
  CC       src/libuv_la-inet.lo
  CC       src/libuv_la-threadpool.lo
  CC       src/libuv_la-uv-common.lo
  CC       src/libuv_la-version.lo
  CC       src/unix/libuv_la-async.lo
  CC       src/unix/libuv_la-core.lo
  CC       src/unix/libuv_la-dl.lo
  CC       src/unix/libuv_la-fs.lo
  CC       src/unix/libuv_la-getaddrinfo.lo
src/unix/fs.c: In function 'uv__fs_mkdtemp':
src/unix/fs.c:262:3: warning: implicit declaration of function 'mkdtemp' [-Wimplicit-function-declaration]
   return mkdtemp((char*) req->path) ? 0 : -1;
   ^
  CC       src/unix/libuv_la-getnameinfo.lo
  CC       src/unix/libuv_la-loop-watcher.lo
  CC       src/unix/libuv_la-loop.lo
  CC       src/unix/libuv_la-pipe.lo
  CC       src/unix/libuv_la-poll.lo
  CC       src/unix/libuv_la-process.lo
  CC       src/unix/libuv_la-signal.lo
  CC       src/unix/libuv_la-stream.lo
  CC       src/unix/libuv_la-tcp.lo
  CC       src/unix/libuv_la-thread.lo
  CC       src/unix/libuv_la-timer.lo
  CC       src/unix/libuv_la-tty.lo
  CC       src/unix/libuv_la-udp.lo
  CC       src/unix/libuv_la-aix.lo
src/unix/aix.c: In function 'uv_uptime':
src/unix/aix.c:940:31: warning: 'boot_time' may be used uninitialized in this function [-Wmaybe-uninitialized]
   *uptime = time(NULL) - boot_time;
                               ^
  CCLD     libuv.la
ld: 0711-317 ERROR: Undefined symbol: .uv__hrtime
ld: 0711-345 Use the -bloadmap or -bnoquiet option to obtain more information.
collect2: error: ld returned 8 exit status
make: 1254-004 The error code from the last command is 1.
```
</details>

**Comment by gireeshpunathil** [comment]
This AIX failure is due to the fact that aix-common.c (that contain the definition for the said symbol) is not in the compilation unit.

```patch
diff --git a/Makefile.am b/Makefile.am
index b94fdd6..56c3fec 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -328,7 +328,7 @@ libuv_la_CFLAGS += -D_ALL_SOURCE \
                    -D_THREAD_SAFE \
                    -DHAVE_SYS_AHAFS_EVPRODS_H
 include_HEADERS += include/uv-aix.h
-libuv_la_SOURCES += src/unix/aix.c
+libuv_la_SOURCES += src/unix/aix.c src/unix/aix-common.c
 endif
```

While this solves it, I guess we need to replicate everything we do in uv.gyp into Makefile.am as well (as they represent two different build methods) - @bnoordhuis I am right?

**Comment by santigimeno** [comment]
>  I guess we need to replicate everything we do in uv.gyp into Makefile.am as well (as they represent two different build methods)

@gireeshpunathil you're right

**Comment by dmabupt** [comment]
@bnoordhuis Since IBM i PASE has not supported autoconf/automake yet, shall I just copy the code change from @gireeshpunathil for now?

**Comment by bnoordhuis** [comment]
CI: https://ci.nodejs.org/job/libuv-test-commit/587/

**Comment by dmabupt** [comment]
Seems the build works fine, Squashed the commits.

**Comment by bnoordhuis** [comment]
@dmabupt Looks good, thanks.  Final question: the Author field of the patch is currently set to "mengxumx".  Do you want me to keep it like that or should I write down your real name?

**Comment by dmabupt** [comment]
@bnoordhuis Thanks for the information. Using my real name "Xu Meng" is better. Thank you!

**Comment by bnoordhuis** [comment]
Landed in bb3d093bb, thanks!
